feature: Add options to ignore imports from tests or other files.

mrseanryan opened the pull request:
feature: Add options to ignore imports from tests or other files.

Addresses part of #2 and fixes #98 

The options are:

| Option name               | Description                                                                                                  | Example                             |
| ------------------------- | ------------------------------------------------------------------------------------------------------------ | ----------------------------------- |
| `ignoreFiles`             | Ignore files with filenames that match the given regex. Use this to exclude groups of files - for example test files and their utilities.                                                                                                                                                                                                              | `--ignoreFiles=\.(spec|test)`         |
| `ignoreProductionFiles`   | Only scan **test** files (so ignore non-test 'production' files).                                                                                                                                                                                                                                                                                      | `--ignoreProductionFiles`             |
| `ignoreTestFiles`         | Only scan **production** files (ignore all test files, like `spec.ts(x)` or `test.ts(x)` or `TestUtils.ts`). Use this to detect production code that is only used in tests (so is dead code). Note: this will NOT detect unused exports in test code - for that, you can run `ts-unused-exports` separately with the `--ignoreProductionFiles` option.                             | `--ignoreTestFiles` |

BREAKING CHANGE:
- rename the option `ignorePaths` to `excludePathsFromReport` - (otherwise is confusing!)

mrseanryan commented:
@pzavolinsky  I think this is good to go.

mrseanryan commented:
@pzavolinsky on 2nd thought:

For 'test utils' code, using this option could cause fake positives. 

I think the option should only be applied, where the exporting file is not also a match. 

So if:

TestUtils.ts `export namespace TestUtils {} `

and x.spec.ts:
```ts
import UsedOnlyInTests from "../main";
import TestUtils from "./TestUtils";
```

Then `UsedOnlyIbTests` ships be flagged, but not `TestUtils`

Note: this requires an option like: 
`--ignoreImportsFrom=(Test|\.test)`

It's difficult to have a sensible default for such cases. 

Maybe the match needs to look at the directory also?

pzavolinsky commented:
Hey @mrseanryan the code looks pretty simple to me.

From a *feature* perspective this feels like it has many corner cases (not the code but what we are trying to achieve).

Some thoughts:
- Exclude files from analysis: using the the approach you already coded, if a file matches an exclusion pattern, we just pretend that file doesn't exist. 
    - Pro: pretty easy to implement.
    - Con: we lose analysis on _those_ files, so if they export unused stuff themselves we won't report it.

- Treat matching files differently: the exclusion patterns divide the codebase in two ("matching" and "not matching"). 
   - Imports from "matching" files do not pin symbols exported by "non-matching" files (e.g. tests do not pin production symbols).
   - Imports from "non-matching" files do not pin symbols exported by "matching" files (if the exclusion criteria is test vs actual code, this means productive code importing testing symbols, doesn't look very probable/correct). 
  - Imports within the same group count as usual.
  - Pro: maybe has better problem fix (although I'm not super sure about it)
  - Con: considerable harder to implement. Might make our codebase harder to understand, maintain, etc.

There are probably other solutions out there that are better than those two. We might also be able to bend the problem a little to get away with something simpler.

What do you think?

pzavolinsky commented:
One more thing. If we had config files, this could be solved by doing the first (simple) solution and having two configs in your project, one for `production` and one for `test`, each with opposite exclusion patterns.

mrseanryan commented:
hi @pzavolinsky 

Yes, I think we need something like "Treat matching files differently" - so having a concept of distinct groups of files.

In theory, we could have an option structure that allows for configuring different groups.

But, yes having config files would be simpler (their folder is a group). 
I guess it's already a convention in git and tsc.

mrseanryan commented:
possible approach:

- try to finish this PR, that introduces the new simpler options (no 'groups')
- later introduce config files - see #99

mrseanryan commented:
hi @pzavolinsky  have updated the PR.

- added more tests, to cover your concept of production vs test files

- changed the options, (see text at the top of this PR).  The previous version of this PR had options that could result in false positives for example with 'test utils' files).

- re config files - it would be idea - but I think for now, a similar effect can be achieved by running `ts-unused-exports` more than once, with different options. for example:

```
./node_modules/.bin/ts-unused-exports ./tsconfig.json --ignoreTestFiles

./node_modules/.bin/ts-unused-exports ./tsconfig.json --ignoreProductionFiles
```

mrseanryan commented:
@pzavolinsky  have rebased this one
